// @livechat/angular-widget/bundles/livechat-angular-widget.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('rxjs')) :
    typeof define === 'function' && define.amd ? define('@livechat/angular-widget', ['exports', '@angular/core', 'rxjs'], factory) :
    (global = global || self, factory((global.livechat = global.livechat || {}, global.livechat['angular-widget'] = {}), global.ng.core, global.rxjs));
}(this, (function (exports, core, rxjs) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */

    var extendStatics = function(d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };

    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    var __assign = function() {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };

    function __rest(s, e) {
        var t = {};
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
        if (s != null && typeof Object.getOwnPropertySymbols === "function")
            for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
                if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                    t[p[i]] = s[p[i]];
            }
        return t;
    }

    function __decorate(decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    }

    function __param(paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    }

    function __metadata(metadataKey, metadataValue) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(metadataKey, metadataValue);
    }

    function __awaiter(thisArg, _arguments, P, generator) {
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
            function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
            function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }

    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f) throw new TypeError("Generator is already executing.");
            while (_) try {
                if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
                if (y = 0, t) op = [op[0] & 2, t.value];
                switch (op[0]) {
                    case 0: case 1: t = op; break;
                    case 4: _.label++; return { value: op[1], done: false };
                    case 5: _.label++; y = op[1]; op = [0]; continue;
                    case 7: op = _.ops.pop(); _.trys.pop(); continue;
                    default:
                        if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                        if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                        if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                        if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                        if (t[2]) _.ops.pop();
                        _.trys.pop(); continue;
                }
                op = body.call(thisArg, _);
            } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
            if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    function __exportStar(m, exports) {
        for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
    }

    function __values(o) {
        var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
        if (m) return m.call(o);
        return {
            next: function () {
                if (o && i >= o.length) o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
    }

    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m) return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
        }
        catch (error) { e = { error: error }; }
        finally {
            try {
                if (r && !r.done && (m = i["return"])) m.call(i);
            }
            finally { if (e) throw e.error; }
        }
        return ar;
    }

    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    function __spreadArrays() {
        for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
        for (var r = Array(s), k = 0, i = 0; i < il; i++)
            for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
                r[k] = a[j];
        return r;
    };

    function __await(v) {
        return this instanceof __await ? (this.v = v, this) : new __await(v);
    }

    function __asyncGenerator(thisArg, _arguments, generator) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var g = generator.apply(thisArg, _arguments || []), i, q = [];
        return i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i;
        function verb(n) { if (g[n]) i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; }
        function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
        function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
        function fulfill(value) { resume("next", value); }
        function reject(value) { resume("throw", value); }
        function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
    }

    function __asyncDelegator(o) {
        var i, p;
        return i = {}, verb("next"), verb("throw", function (e) { throw e; }), verb("return"), i[Symbol.iterator] = function () { return this; }, i;
        function verb(n, f) { i[n] = o[n] ? function (v) { return (p = !p) ? { value: __await(o[n](v)), done: n === "return" } : f ? f(v) : v; } : f; }
    }

    function __asyncValues(o) {
        if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
        var m = o[Symbol.asyncIterator], i;
        return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
        function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
        function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
    }

    function __makeTemplateObject(cooked, raw) {
        if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
        return cooked;
    };

    function __importStar(mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
        result.default = mod;
        return result;
    }

    function __importDefault(mod) {
        return (mod && mod.__esModule) ? mod : { default: mod };
    }

    var LivechatWidgetComponent = /** @class */ (function () {
        function LivechatWidgetComponent() {
            this.onChatLoaded = new rxjs.Subject();
            this.onAfterLoad = new rxjs.Subject();
            this.onBeforeLoad = new rxjs.Subject();
            this.onChatEnded = new rxjs.Subject();
            this.onChatStarted = new rxjs.Subject();
            this.onChatStateChanged = new rxjs.Subject();
            this.onChatWindowHidden = new rxjs.Subject();
            this.onChatWindowMinimized = new rxjs.Subject();
            this.onChatWindowOpened = new rxjs.Subject();
            this.onMessage = new rxjs.Subject();
            this.onPostchatSurveySubmitted = new rxjs.Subject();
            this.onPrechatSurveySubmitted = new rxjs.Subject();
            this.onRatingCommentSubmitted = new rxjs.Subject();
            this.onRatingSubmitted = new rxjs.Subject();
            this.onTicketCreated = new rxjs.Subject();
            this.onWidgetResize = new rxjs.Subject();
        }
        LivechatWidgetComponent.prototype.ngOnInit = function () {
            var _this = this;
            if (!window['LC_API']) {
                window['__lc'] = window['__lc'] || {};
                window['__lc'].license = this.licenseId;
                window['__lc'].group = this.group;
                window['__lc'].chat_between_groups = this.chatBetweenGroups;
                window['__lc'].params = this.params;
                window['__lc'].visitor = this.visitor;
                window['__lc'].ga_version = this.gaVersion;
                var api_1 = { _q: [], _h: null, call: function () {
                        var args = [];
                        for (var _i = 0; _i < arguments.length; _i++) {
                            args[_i] = arguments[_i];
                        }
                        return api_1._h && api_1._h.apply(null, ['call', [].slice.call(args)]);
                    } };
                window['LiveChatWidget'] = api_1;
                this.lc = document.createElement('script');
                this.lc.type = 'text/javascript';
                this.lc.async = true;
                this.lc.src = 'https://cdn.livechatinc.com/tracking.js';
                var s = document.getElementsByTagName('script')[0];
                s.parentNode.insertBefore(this.lc, s);
                this.lc.addEventListener('load', function () {
                    _this._chatLoaded(window['LC_API']);
                });
                this.lc.addEventListener('error', function (error) {
                    console.error(error);
                });
            }
        };
        LivechatWidgetComponent.prototype.ngOnDestroy = function () {
            if (window['LiveChatWidget'] && typeof window['LiveChatWidget'].call === 'function') {
                window['LiveChatWidget'].call('destroy');
                this.lc && this.lc.remove();
                ['secure.livechat.com', 'secure.livechatinc.com'].forEach(function (domain) {
                    Array.from(document.querySelectorAll("[src*=\"" + domain + "\"]")).forEach(function (s) { return s.remove(); });
                });
            }
        };
        LivechatWidgetComponent.prototype._chatLoaded = function (LC_API) {
            this._setCallbacks(LC_API);
            this._setMethods(LC_API);
            this.onChatLoaded.next(LC_API);
        };
        LivechatWidgetComponent.prototype._setCallbacks = function (LC_API) {
            var _this = this;
            LC_API.on_after_load = function () { return _this.onAfterLoad.next(); };
            LC_API.on_before_load = function () { return _this.onBeforeLoad.next(); };
            LC_API.on_chat_ended = function () { return _this.onChatEnded.next(); };
            LC_API.on_chat_started = function (data) { return _this.onChatStarted.next(data); };
            LC_API.on_chat_state_changed = function (data) { return _this.onChatStateChanged.next(data); };
            LC_API.on_chat_window_hidden = function () { return _this.onChatWindowHidden.next(); };
            LC_API.on_chat_window_minimized = function () { return _this.onChatWindowMinimized.next(); };
            LC_API.on_chat_window_opened = function () { return _this.onChatWindowOpened.next(); };
            LC_API.on_message = function (data) { return _this.onMessage.next(data); };
            LC_API.on_postchat_survey_submitted = function (data) { return _this.onPostchatSurveySubmitted.next(data); };
            LC_API.on_prechat_survey_submitted = function (data) { return _this.onPrechatSurveySubmitted.next(data); };
            LC_API.on_rating_comment_submitted = function (data) { return _this.onRatingCommentSubmitted.next(data); };
            LC_API.on_rating_submitted = function (data) { return _this.onRatingSubmitted.next(data); };
            LC_API.on_ticket_created = function (data) { return _this.onTicketCreated.next(data); };
        };
        LivechatWidgetComponent.prototype._setMethods = function (LC_API) {
            this.agentsAreAvailable = function () { return LC_API.agents_are_available(); };
            this.chatRunning = function () { return LC_API.chat_running(); };
            this.chatWindowHidden = function () { return LC_API.chat_window_hidden(); };
            this.chatWindowMaximized = function () { return LC_API.chat_window_maximized(); };
            this.chatWindowMinimized = function () { return LC_API.chat_window_minimized(); };
            this.closeChat = function () { return LC_API.close_chat(); };
            this.getChatId = function () { return LC_API.get_chat_id(); };
            this.getChatsNumber = function () { return LC_API.get_chats_number(); };
            this.getLastVisitTimestamp = function () { return LC_API.get_last_visit_timestamp(); };
            this.getPageViewsNumber = function () { return LC_API.get_page_views_number(); };
            this.getVisitorId = function () { return LC_API.get_visitor_id(); };
            this.getVisitsNumber = function () { return LC_API.get_visits_number(); };
            this.hideChatWindow = function () { return LC_API.hide_chat_window(); };
            this.minimizeChatWindow = function () { return LC_API.minimize_chat_window(); };
            this.openChatWindow = function () { return LC_API.open_chat_window(); };
            this.setCustomVariables = function (t) { return LC_API.set_custom_variables(t); };
            this.setVisitorEmail = function (t) { return LC_API.set_visitor_email(t); };
            this.setVisitorName = function (t) { return LC_API.set_visitor_name(t); };
            this.triggerSalesTracker = function (e, t) { return LC_API.trigger_sales_tracker(e, t); };
            this.updateCustomVariables = function (t) { return LC_API.update_custom_variables(t); };
            this.visitorEngaged = function () { return LC_API.visitor_engaged(); };
            this.visitorQueued = function () { return LC_API.visitor_queued(); };
        };
        __decorate([
            core.Input()
        ], LivechatWidgetComponent.prototype, "licenseId", void 0);
        __decorate([
            core.Input()
        ], LivechatWidgetComponent.prototype, "group", void 0);
        __decorate([
            core.Input()
        ], LivechatWidgetComponent.prototype, "chatBetweenGroups", void 0);
        __decorate([
            core.Input()
        ], LivechatWidgetComponent.prototype, "params", void 0);
        __decorate([
            core.Input()
        ], LivechatWidgetComponent.prototype, "visitor", void 0);
        __decorate([
            core.Input()
        ], LivechatWidgetComponent.prototype, "gaVersion", void 0);
        __decorate([
            core.Output()
        ], LivechatWidgetComponent.prototype, "onChatLoaded", void 0);
        __decorate([
            core.Output()
        ], LivechatWidgetComponent.prototype, "onAfterLoad", void 0);
        __decorate([
            core.Output()
        ], LivechatWidgetComponent.prototype, "onBeforeLoad", void 0);
        __decorate([
            core.Output()
        ], LivechatWidgetComponent.prototype, "onChatEnded", void 0);
        __decorate([
            core.Output()
        ], LivechatWidgetComponent.prototype, "onChatStarted", void 0);
        __decorate([
            core.Output()
        ], LivechatWidgetComponent.prototype, "onChatStateChanged", void 0);
        __decorate([
            core.Output()
        ], LivechatWidgetComponent.prototype, "onChatWindowHidden", void 0);
        __decorate([
            core.Output()
        ], LivechatWidgetComponent.prototype, "onChatWindowMinimized", void 0);
        __decorate([
            core.Output()
        ], LivechatWidgetComponent.prototype, "onChatWindowOpened", void 0);
        __decorate([
            core.Output()
        ], LivechatWidgetComponent.prototype, "onMessage", void 0);
        __decorate([
            core.Output()
        ], LivechatWidgetComponent.prototype, "onPostchatSurveySubmitted", void 0);
        __decorate([
            core.Output()
        ], LivechatWidgetComponent.prototype, "onPrechatSurveySubmitted", void 0);
        __decorate([
            core.Output()
        ], LivechatWidgetComponent.prototype, "onRatingCommentSubmitted", void 0);
        __decorate([
            core.Output()
        ], LivechatWidgetComponent.prototype, "onRatingSubmitted", void 0);
        __decorate([
            core.Output()
        ], LivechatWidgetComponent.prototype, "onTicketCreated", void 0);
        __decorate([
            core.Output()
        ], LivechatWidgetComponent.prototype, "onWidgetResize", void 0);
        LivechatWidgetComponent = __decorate([
            core.Component({
                selector: 'livechat-widget',
                template: ""
            })
        ], LivechatWidgetComponent);
        return LivechatWidgetComponent;
    }());

    var LivechatWidgetModule = /** @class */ (function () {
        function LivechatWidgetModule() {
        }
        LivechatWidgetModule = __decorate([
            core.NgModule({
                imports: [],
                declarations: [LivechatWidgetComponent],
                exports: [LivechatWidgetComponent]
            })
        ], LivechatWidgetModule);
        return LivechatWidgetModule;
    }());

    exports.LivechatWidgetComponent = LivechatWidgetComponent;
    exports.LivechatWidgetModule = LivechatWidgetModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
